rep-2001.rst

```diff
@@ -773,6 +775,33 @@ Desktop Full
       packages: [perception, simulation, ros_ign_gazebo_demos]
 
 
+ROS Distribution Independent (September 2022 - Ongoing)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+These variants are not released as ROS packages within each distribution as the others are.
+Instead they are made available as independent packages outside of any ROS distribution.
```

Comment:
I have two comments here which are related.

First of all, one outcome I would love is from this work is for us to be able to simplify the instructions in https://docs.ros.org/en/rolling/Installation/Alternatives/Ubuntu-Development-Setup.html .  In particular, I would like it if https://docs.ros.org/en/rolling/Installation/Alternatives/Ubuntu-Development-Setup.html#install-development-tools-and-ros-tools was reduced to `sudo apt-update && sudo apt install -y ros-dev-tools`.

To get there, we would need to add all of the development tools from that documentation page to the `ros-dev-tools` variant.  Do we want to do that here?  Alternatively, we could add most of those tools as `package.xml` dependencies...somewhere, though it isn't totally clear to me where.

Depending on the answer to that question, we may have to scope the `ros-dev-tools` to be Ubuntu 22.04 and forwards.  That's because many of the packages in that documentation list were not available until Jammy.

Reply:
I get into this in a separate line comment https://github.com/ros-infrastructure/rep/pull/355#discussion_r976894371

Going into each of the packages in that section.


* build-essential: We have to hash out some cross-platform specifics but this is in ros-dev-tools via ros-build-essential.
* cmake: in ros-dev-tools via ros-build-essential
* git: in ros-dev-tools via ros-build-essential (git is a requirement of several vendor packages but we also encourage explicit declaration where it is used so perhaps this could move to dev-tools directly)
* python3-colcon-common-extensions: ros-dev-tools
* flake8 + plugins: We should declare these in ament_flake8, it's been on my wishlist ever since we got them packaged in Ubuntu it just hasn't been my top priority:
  * python3-flake8
  * python3-flake8-blind-except
  * python3-flake8-builtins
  * python3-flake8-class-newline
  * python3-flake8-comprehensions
  * python3-flake8-deprecated
  * python3-flake8-docstrings
  * python3-flake8-import-order
  * python3-flake8-quotes
* python3-pip: I think we can actually drop this entirely, it used to be for installing flake8 extensions from pip but we install everything via apt now. "pip install" does not appear anywhere in that doc. 
* pytest and plugins: These are dependencies of [colcon-core](https://github.com/colcon/colcon-core/blob/0bf904e205aaae3055f9b8fefbdd4f52806b66e6/setup.cfg#L33-L39) although it's noted that they are technically conditional. I think their conditionality may have been due to the fact that they were not all available via apt. But either way colcon requires them to run tests. I could see these being "build essential" by virtue of being "test essential" but it sort of comes down to taste and convention. We expect every package to declare a dependency on its _build_ framework (`cmake` for `cmake`, `ament_cmake` for `ament_cmake`, etc) why not make them declare their test framework as well. The `ament_python` build type used to have a corresponding library but issues with how dependencies are specified in python and what is available during a sourcepkg job caused us to remove that (although new developments have made it potentially possible again in the future). I would expect such a build type to require pytest but as that's a hypothetical I'm not sure what the best thing to do today is yet. It irks me to think that these packages would be required here simply because they don't fit anywhere else.
  * python3-pytest
  * python3-pytest-cov
  * python3-pytest-repeat
  * python3-pytest-rerunfailures
* python3-rosdep: ros-dev-tools
* python3-setuptools: ros-build-essential
* python3-vcstool: ros-dev-tools
* wget: hah, this one I thought I actually tried to remove in favor of curl. It's in the instructions in order to fetch the ros2.repos file from GitHub. I am not sure that swapping wget for curl makes me feel any better but it is what we should do.

Reply:
One overall comment that I just realized; all of these package names should be `rosdep` names, not Ubuntu package names.  Otherwise we can never make this work for other operating systems.  With that said, here are some comments on the package list (if I agree with your statement, I've just not said anything for brevity):

> build-essential: We have to hash out some cross-platform specifics but this is in ros-dev-tools via ros-build-essential.

But it is currently not explicitly listed in `ros-build-essential`.  Also, there is no rosdep key for it.

> flake8 + plugins: We should declare these in ament_flake8, it's been on my wishlist ever since we got them packaged in Ubuntu it just hasn't been my top priority

All right, that is fair enough.

>  python3-pip: I think we can actually drop this entirely, it used to be for installing flake8 extensions from pip but we install everything via apt now. "pip install" does not appear anywhere in that doc.

Good point.  I'm fine with dropping it.

> pytest and plugins: ... why not make them declare their test framework as well

I'm not sure I would classify the use of those plugins as a "test framework".  Pytest is the test framework, and most (if not all) packages that use it have an explicit dependency in their `package.xml` on the pytest rosdep key.  Some of them also declare dependencies on other `pytest` extensions, though that is not universal.  The issue here is that typically the package itself doesn't depend on `rerunfailures` and such; those are requirements imposed upon it by the external tool (usually `colcon`).  So this is a long way of saying that I think those plugins should be dependencies of `colcon-core` (which they currently are), and we should not add them here. (incidentally, `python3-pytest-cov` is a `Recommend` for `colcon-core`, not a straight-up dependency.  Maybe we should change that)

> wget: hah, this one I thought I actually tried to remove in favor of curl. It's in the instructions in order to fetch the ros2.repos file from GitHub. I am not sure that swapping wget for curl makes me feel any better but it is what we should do.

Actually, `wget` isn't used for the default instructions anymore; we switched to having `vcstool` directly get the file over HTTP (since it can do that).  That said, we use `wget` in enough places in the documentation (I count 47 uses) that I think we should add it here, even though it largely duplicates functionality with curl.

Reply:
> Actually, `wget` isn't used for the default instructions anymore; we switched to having `vcstool` directly get the file over HTTP (since it can do that). That said, we use `wget` in enough places in the documentation (I count 47 uses) that I think we should add it here, even though it largely duplicates functionality with curl.

😖

> One overall comment that I just realized; all of these package names should be `rosdep` names, not Ubuntu package names.

The rosdep names for all but the most venerable of keys should match the Ubuntu package names so this is probably not a significant issue.

Reply:
Sounds good.  This generally looks good to me now.